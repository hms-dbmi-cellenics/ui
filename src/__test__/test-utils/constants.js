export default {
  EXPERIMENT_ID: 'testae48e318dab9a1bd0bexperiment',
  CLONED_EXPERIMENT_ID: 'testae48e318dab9a1bd0bexperclone',
  PROJECT_WITH_SAMPLES_UUID: '380c53b4-mock-test-8091-4c82336d6d49',
  PROJECT_WITHOUT_SAMPLE_UUID: '4761594b-mock-test-ba89-c2c812d39bb5',
  EXPERIMENT_NAME: 'Test Experiment',
  PROJECT_ID: 'test24b6-8600-test-mock-63822d2fmock',
  SAMPLE_ID: 'test9188-d682-test-mock-cb6d644cmock',
  API_ENDPOINT: /^http?:\/\/localhost:3000.*$/,
  S3_ENDPOINT: 'http://mock.s3.amazonaws.com',
  // pragma: allowlist secret
  JWT_TOKEN: 'Bearer eyJhbGciOiJIUzI1NiIsInR5cCI6Ikpfake.eyJzdWIiOiIxMjM0NTY3ODkwIiwifakeZSI6IkpvaG4gRG9lIiwiaWF0IjoxNTfakeM5MDIyfQ.fakexwRfakeKKF2QT4fwpMeJf36POk6yJV_adQsfake',
  MOCK_DATETIME: '0000-00-00T00:00:00.000Z',
  ETAG: '380c53b4-mock-etag-8091-4c82336d6d49',
};
