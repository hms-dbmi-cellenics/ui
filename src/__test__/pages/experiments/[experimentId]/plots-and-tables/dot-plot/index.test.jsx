import React from 'react';
import _ from 'lodash';

import { act } from 'react-dom/test-utils';
import { render, screen, fireEvent } from '@testing-library/react';
import { mount } from 'enzyme';
import { within } from '@testing-library/dom';
import '@testing-library/jest-dom';
import { Provider } from 'react-redux';

import fetchMock, { enableFetchMocks } from 'jest-fetch-mock';

import fake from '__test__/test-utils/constants';
import mockAPI, {
  generateDefaultMockAPIResponses,
  promiseResponse,
  statusResponse,
  delayedResponse,
} from '__test__/test-utils/mockAPI';

import { seekFromS3 } from 'utils/work/seekWorkResponse';

import createTestComponentFactory from '__test__/test-utils/testComponentFactory';
import { makeStore } from 'redux/store';

import { loadBackendStatus } from 'redux/actions/backendStatus';
import DotPlotPage from 'pages/experiments/[experimentId]/plots-and-tables/dot-plot/index';
import {
  EXPERIMENT_SETTINGS_INFO_UPDATE,
} from 'redux/actionTypes/experimentSettings';
import paginatedGeneExpressionData from '__test__/data/paginated_gene_expression.json';
import cellSetsDataWithScratchpad from '__test__/data/cell_sets_with_scratchpad.json';
import dotPlotData from '__test__/data/dotplot_plotdata.json';
import userEvent from '@testing-library/user-event';
import { plotNames } from 'utils/constants';
import ExportAsCSV from 'components/plots/ExportAsCSV';

import waitForComponentToPaint from '__test__/test-utils/waitForComponentToPaint';
import { arrayMoveImmutable } from 'utils/array-move';

jest.mock('components/plots/ExportAsCSV', () => jest.fn(() => (<></>)));
jest.mock('components/header/UserButton', () => () => <></>);
jest.mock('react-resize-detector', () => (props) => {
  // eslint-disable-next-line react/prop-types
  const { children } = props;
  return children({ width: 800, height: 800 });
});

jest.mock('object-hash', () => {
  const objectHash = jest.requireActual('object-hash');
  const mockWorkResultETag = jest.requireActual('__test__/test-utils/mockWorkResultETag').default;

  const mockWorkRequestETag = (ETagParams) => {
    if (ETagParams.body.name === 'ListGenes') return 'paginated-gene-expression';
    if (ETagParams.body.name === 'DotPlot') return 'dot-plot-data';
  };
  const mockGeneExpressionETag = (ETagParams) => `${ETagParams.missingGenesBody.genes.join('-')}-expression`;

  return mockWorkResultETag(objectHash, mockWorkRequestETag, mockGeneExpressionETag);
});

jest.mock('utils/work/seekWorkResponse', () => ({
  __esModule: true,
  dispatchWorkRequest: jest.fn(() => true),
  seekFromS3: jest.fn(),
}));

const mockWorkerResponses = {
  'paginated-gene-expression': () => paginatedGeneExpressionData,
  'dot-plot-data': () => dotPlotData,
};

const experimentId = fake.EXPERIMENT_ID;
const plotUuid = 'dotPlotMain';

const customAPIResponses = {
  [`experiments/${experimentId}/cellSets`]: () => promiseResponse(
    JSON.stringify(cellSetsDataWithScratchpad),
  ),
  [`/plots/${plotUuid}`]: (req) => {
    if (req.method === 'PUT') return promiseResponse(JSON.stringify('OK'));
    return statusResponse(404, 'Not Found');
  },
};

const mockAPIResponse = _.merge(
  generateDefaultMockAPIResponses(experimentId),
  customAPIResponses,
);

const defaultProps = { experimentId };
const dotPlotPageFactory = createTestComponentFactory(DotPlotPage, defaultProps);

// Helper function to get genes held within the tree
const getTreeGenes = (container) => {
  const treeNodeList = container.querySelectorAll('span[class*=ant-tree-title]');
  return Array.from(treeNodeList).map((node) => node.textContent);
};

// Helper function to get current order of displayed genes in enzyme tests
const getCurrentGeneOrder = (component) => {
  const treeNodes = component.find('div.ant-tree-treenode');
  const newOrder = [];
  treeNodes.forEach((node) => {
    newOrder.push(node.text());
  });
  newOrder.splice(0, 1);
  return newOrder;
};

const renderDotPlot = async (store) => {
  await act(async () => {
    render(
      <Provider store={store}>
        {dotPlotPageFactory()}
      </Provider>,
    );
  });
};

const renderDotPlotForEnzyme = async (store) => (
  mount(
    <Provider store={store}>
      {dotPlotPageFactory()}
    </Provider>,
  )
);

enableFetchMocks();

let storeState = null;

describe('Dot plot page', () => {
  beforeEach(async () => {
    jest.clearAllMocks();

    seekFromS3
      .mockReset()
      // 1st call to list genes
      .mockImplementationOnce(() => null)
      .mockImplementationOnce((Etag) => mockWorkerResponses[Etag]())
      // 2nd call to paginated gene expression
      .mockImplementationOnce(() => null)
      .mockImplementationOnce((Etag) => mockWorkerResponses[Etag]());

    fetchMock.resetMocks();
    fetchMock.mockIf(/.*/, mockAPI(mockAPIResponse));

    storeState = makeStore();

    await storeState.dispatch(loadBackendStatus(experimentId));

    storeState.dispatch({
      type: EXPERIMENT_SETTINGS_INFO_UPDATE,
      payload: {
        experimentId: fake.EXPERIMENT_ID,
        experimentName: fake.EXPERIMENT_NAME,
      },
    });
  });

  it('Renders the plot page correctly', async () => {
    await renderDotPlot(storeState);

    // There is the text Dot plot show in the breadcrumbs
    expect(screen.getByText(new RegExp(plotNames.DOT_PLOT, 'i'))).toBeInTheDocument();

    // It has the required dropdown options
    expect(screen.getByText(/Gene selection/i)).toBeInTheDocument();
    expect(screen.getByText(/Select data/i)).toBeInTheDocument();
    expect(screen.getByText(/Size scale/i)).toBeInTheDocument();
    expect(screen.getByText(/Main schema/i)).toBeInTheDocument();
    expect(screen.getByText(/Axes and margins/i)).toBeInTheDocument();
    expect(screen.getByText(/Colours/i)).toBeInTheDocument();
    expect(screen.getByText(/Legend/i)).toBeInTheDocument();

    // It shows the plot
    expect(screen.getByRole('graphics-document', { name: 'Vega visualization' })).toBeInTheDocument();

    // csv data is passed correctly
    expect(ExportAsCSV.mock.calls).toMatchSnapshot();
  });

  it('Shows a skeleton if config is not loaded', async () => {
    const noConfigResponse = {
      ...mockAPIResponse,
      [`/plots/${plotUuid}`]: () => delayedResponse({ body: 'Not found', status: 404 }),
    };

    fetchMock.mockIf(/.*/, mockAPI(noConfigResponse));

    await renderDotPlot(storeState);

    expect(screen.getByRole('list')).toHaveClass('ant-skeleton-paragraph');
  });

  it('Shows an error if there are errors loading cell sets', async () => {
    const cellSetsErrorResponse = {
      ...mockAPIResponse,
      [`experiments/${experimentId}/cellSets`]: () => statusResponse(404, 'Nothing found'),
    };

    fetchMock.mockIf(/.*/, mockAPI(cellSetsErrorResponse));

    await renderDotPlot(storeState);

    expect(screen.getByText(/Error loading cell sets/i)).toBeInTheDocument();
  });

  it('Shows platform error if there are errors fetching the work', async () => {
    const errorResponse = {
      ...mockWorkerResponses,
      'dot-plot-data': () => { throw new Error('error'); },
    };

    seekFromS3
      .mockReset()
      .mockImplementationOnce(() => null)
      .mockImplementationOnce((Etag) => errorResponse[Etag]())
      .mockImplementationOnce(() => null)
      .mockImplementationOnce((Etag) => errorResponse[Etag]());

    await renderDotPlot(storeState);

    expect(screen.getByText(/Error loading plot data/i)).toBeInTheDocument();
    expect(screen.getByText(/Check the options that you have selected and try again/i)).toBeInTheDocument();
  });

  it('Shows an empty message if there is no data to show in the plot', async () => {
    const emptyResponse = {
      ...mockWorkerResponses,
      'dot-plot-data': () => [],
    };

    seekFromS3
      .mockReset()
      .mockImplementationOnce(() => null)
      .mockImplementationOnce((Etag) => emptyResponse[Etag]())
      .mockImplementationOnce(() => null)
      .mockImplementationOnce((Etag) => emptyResponse[Etag]());

    await renderDotPlot(storeState);

    expect(screen.getByText(/There is no data to show/i)).toBeInTheDocument();
    expect(screen.getByText(/Select another option from the 'Select data' menu/i)).toBeInTheDocument();
  });

  it('Should show a no data error if user is using marker gene and selected filter sets are not represented in more than 1 group in the base cell set', async () => {
    await renderDotPlot(storeState);

    seekFromS3
      .mockReset()
      // 1st call to list genes
      .mockImplementationOnce(() => null)
      .mockImplementationOnce((Etag) => mockWorkerResponses[Etag]())
      // 2nd call to load dot plot
      .mockImplementationOnce(() => null)
      .mockImplementationOnce((Etag) => mockWorkerResponses[Etag]())
      // 3rd call to load dot plot
      .mockImplementationOnce(() => null)
      .mockImplementationOnce((Etag) => mockWorkerResponses[Etag]());

    // Use marker genes
    await act(async () => {
      userEvent.click(screen.getByText(/Marker genes/i));
    });

    // Select data
    await act(async () => {
      userEvent.click(screen.getByText(/Select data/i));
    });

    // Select samples
    const selectBaseCells = screen.getByRole('combobox', { name: 'selectCellSets' });

    await act(async () => {
      fireEvent.change(selectBaseCells, { target: { value: 'Samples' } });
    });

    const baseOption = screen.getByText(/Samples/);

    await act(async () => {
      fireEvent.click(baseOption);
    });

    // Select the filter sets
    const selectFilterCells = screen.getByRole('combobox', { name: 'selectPoints' });

    await act(async () => {
      fireEvent.change(selectFilterCells, { target: { value: 'Samples' } });
    });

    const filterOption = screen.getByText(/Copied WT2/);

    await act(async () => {
      fireEvent.click(filterOption);
    });

    expect(screen.getByText(/There is no data to show/i)).toBeInTheDocument();
    expect(screen.getByText(/The cell set that you have chosen to display is repesented by only one group/i)).toBeInTheDocument();
    expect(screen.getByText(/A comparison can not be run to determine the top marker genes/i)).toBeInTheDocument();
    expect(screen.getByText(/Select another option from the 'Select data' menu/i)).toBeInTheDocument();
  });

  it('removing a gene keeps the order', async () => {
    await renderDotPlot(storeState);

    const geneTree = screen.getByRole('tree');

    // first three genes of the data should be loaded by default
    const loadedGenes = paginatedGeneExpressionData.rows.map((row) => (row.gene_names)).slice(0, 3);

    // The genes in Data 5 should be in the tree
    loadedGenes.forEach((geneName) => {
      expect(within(geneTree).getByText(geneName)).toBeInTheDocument();
    });

    // Remove a gene using the X button
    const genesListBeforeRemoval = getTreeGenes(geneTree);

    const geneToRemove = within(geneTree).getByText(genesListBeforeRemoval[1]);

    const geneRemoveButton = geneToRemove.nextSibling.firstChild;

    userEvent.click(geneRemoveButton);

    const genesListAfterRemoval = getTreeGenes(geneTree);

    // remove element from list manually to compare
    genesListBeforeRemoval.splice(1, 1);

    // The gene should be deleted from the list
    expect(_.isEqual(genesListAfterRemoval, genesListBeforeRemoval)).toEqual(true);
  });

  it('searches for genes and adds a valid gene', async () => {
    await renderDotPlot(storeState);

    const geneTree = screen.getByRole('tree');
    const initialOrder = getTreeGenes(geneTree);

    // check placeholder text is loaded
    expect(screen.getByText('Search for genes...')).toBeInTheDocument();

    const searchBox = screen.getByRole('combobox');

    // search for genes using lowercase
    userEvent.type(searchBox, 'ap');

    // antd creates multiple elements for options
    // find option element by title, clicking on element with role='option' does nothing
    const option = screen.getByTitle('Apoe');

    await act(async () => {
      // the element has pointer-events set to 'none', skip check
      // based on https://stackoverflow.com/questions/61080116
      userEvent.click(option, undefined, { skipPointerEventsCheck: true });
    });

    // check the search text is modified after selecting a valid option
    expect(searchBox.value).toBe('Apoe, ');

    const geneAddButton = screen.getByText('Add');

    userEvent.click(geneAddButton);

    // check the selected gene was added
    expect(within(geneTree).getByText('Apoe')).toBeInTheDocument();

    // check the genes were not re-ordered when adding
    initialOrder.push('Apoe');
    expect(_.isEqual(initialOrder, getTreeGenes(geneTree))).toEqual(true);
  });

  it('tries to select an already loaded gene and clears the input', async () => {
    await renderDotPlot(storeState);

    const searchBox = screen.getByRole('combobox');

    userEvent.type(searchBox, 'ly');

    const option = screen.getByTitle('Lyz2');

    // expecting option to be disabled throws error, click the option instead and check reaction
    await act(async () => {
      userEvent.click(option, undefined, { skipPointerEventsCheck: true });
    });

    // search box shouldn't clear when selecting an already loaded gene
    expect(searchBox.value).toBe('ly');

    // clear button is automatically generated by antd and cannot be easily accessed
    const clearButton = searchBox.closest('div[class*=ant-select-auto-complete]').lastChild;

    userEvent.click(clearButton);

    expect(searchBox.value).toBe('');
  });

  it('resets the data', async () => {
    await renderDotPlot(storeState);

    seekFromS3
      .mockReset()
      // 1st call to list genes
      .mockImplementationOnce(() => null)
      .mockImplementationOnce((Etag) => mockWorkerResponses[Etag]())
      // 2nd call to load dot plot
      .mockImplementationOnce(() => null)
      .mockImplementationOnce((Etag) => mockWorkerResponses[Etag]())
      // 3rd call to load dot plot
      .mockImplementationOnce(() => null)
      .mockImplementationOnce((Etag) => mockWorkerResponses[Etag]());

    // add a gene to prepare for reset
    const searchBox = screen.getByRole('combobox');

    userEvent.type(searchBox, 'ap');

    const option = screen.getByTitle('Apoe');

    await act(async () => {
      userEvent.click(option, undefined, { skipPointerEventsCheck: true });
    });

    const resetButton = screen.getAllByText('Reset')[1];

    await act(async () => {
      userEvent.click(resetButton);
    });

    // expect the gene only within the options of the search box, antd creates 2 elements
    expect(screen.getAllByText('Apoe').length).toBe(2);
  });
});

// drag and drop is impossible in RTL, use enzyme
describe('Drag and drop enzyme tests', () => {
  let component;
  let tree;
  let loadedGenes;

  beforeEach(async () => {
    jest.clearAllMocks();

    seekFromS3
      .mockReset()
      // 1st call to list genes
      .mockImplementationOnce(() => null)
      .mockImplementationOnce((Etag) => mockWorkerResponses[Etag]())
      // 2nd call to paginated gene expression
      .mockImplementationOnce(() => null)
      .mockImplementationOnce((Etag) => mockWorkerResponses[Etag]());

    fetchMock.resetMocks();
    fetchMock.mockIf(/.*/, mockAPI(mockAPIResponse));

    storeState = makeStore();

    await storeState.dispatch(loadBackendStatus(experimentId));

    storeState.dispatch({
      type: EXPERIMENT_SETTINGS_INFO_UPDATE,
      payload: {
        experimentId: fake.EXPERIMENT_ID,
        experimentName: fake.EXPERIMENT_NAME,
      },
    });

    component = await renderDotPlotForEnzyme(storeState);

    await waitForComponentToPaint(component);

    component.update();

    // antd renders 5 elements, use the first one
    tree = component.find({ 'data-testid': 'HierachicalTreeGenes' }).at(0);
    loadedGenes = paginatedGeneExpressionData.rows.map((row) => (row.gene_names)).slice(0, 3).reverse();
  });

  it('changes nothing on drop in place', async () => {
    // default genes are in the tree
    loadedGenes.forEach((geneName) => {
      expect(tree.containsMatchingElement(geneName));
    });

    // dropping in place does nothing
    const info = {
      dragNode: { key: 1, pos: '0-1' },
      dropPosition: 1,
      node: { dragOver: false },
    };

    tree.getElement().props.onDrop(info);

    await act(async () => {
      component.update();
    });

    const newOrder = getCurrentGeneOrder(component);

    expect(_.isEqual(newOrder, loadedGenes)).toEqual(true);
  });

  it('re-orders genes correctly', async () => {
    // default genes are in the tree
    loadedGenes.forEach((geneName) => {
      expect(tree.containsMatchingElement(geneName));
    });
    // dropping to gap re-orders genes
    const info = {
      dragNode: { key: 0, pos: '0-0' },
      dropPosition: 2,
      node: { dragOver: false },
    };

    tree.getElement().props.onDrop(info);

    await act(async () => {
      component.update();
    });

    const newOrder = getCurrentGeneOrder(component);

    const expectedOrder = arrayMoveImmutable(loadedGenes, 0, 1);

    expect(_.isEqual(newOrder, expectedOrder)).toEqual(true);
  });
});
