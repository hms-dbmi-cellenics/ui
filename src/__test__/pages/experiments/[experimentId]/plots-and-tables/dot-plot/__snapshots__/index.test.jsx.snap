// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Dot plot page Renders the plot page correctly 1`] = `
[
  {
    "data": [
      {
        "Average expression": 0.0417,
        "Cluster": "Cluster 5",
        "Fraction of cells expressing gene": 1.4031,
        "Gene": "Apoe",
      },
      {
        "Average expression": 0.0273,
        "Cluster": "Cluster 6",
        "Fraction of cells expressing gene": 0.6678,
        "Gene": "Apoe",
      },
      {
        "Average expression": 0.0276,
        "Cluster": "Cluster 7",
        "Fraction of cells expressing gene": 0.6211,
        "Gene": "Apoe",
      },
      {
        "Average expression": 0.0254,
        "Cluster": "Cluster 8",
        "Fraction of cells expressing gene": 2.4631,
        "Gene": "Apoe",
      },
      {
        "Average expression": 3.7359,
        "Cluster": "Cluster 9",
        "Fraction of cells expressing gene": 32.716,
        "Gene": "Apoe",
      },
      {
        "Average expression": 0,
        "Cluster": "Cluster 10",
        "Fraction of cells expressing gene": 0,
        "Gene": "Apoe",
      },
      {
        "Average expression": 95.087,
        "Cluster": "Cluster 11",
        "Fraction of cells expressing gene": 94.6429,
        "Gene": "Apoe",
      },
      {
        "Average expression": 75.4928,
        "Cluster": "Cluster 12",
        "Fraction of cells expressing gene": 63.6364,
        "Gene": "Apoe",
      },
      {
        "Average expression": 5.2359,
        "Cluster": "Cluster 13",
        "Fraction of cells expressing gene": 19.0476,
        "Gene": "Apoe",
      },
      {
        "Average expression": 0.0246,
        "Cluster": "Cluster 14",
        "Fraction of cells expressing gene": 1.9608,
        "Gene": "Apoe",
      },
      {
        "Average expression": 0.0251,
        "Cluster": "Cluster 0",
        "Fraction of cells expressing gene": 0.8608,
        "Gene": "Apoe",
      },
      {
        "Average expression": 0.0102,
        "Cluster": "Cluster 1",
        "Fraction of cells expressing gene": 0.3534,
        "Gene": "Apoe",
      },
      {
        "Average expression": 0.024,
        "Cluster": "Cluster 2",
        "Fraction of cells expressing gene": 0.5848,
        "Gene": "Apoe",
      },
      {
        "Average expression": 0.0283,
        "Cluster": "Cluster 3",
        "Fraction of cells expressing gene": 0.7857,
        "Gene": "Apoe",
      },
      {
        "Average expression": 0.0213,
        "Cluster": "Cluster 4",
        "Fraction of cells expressing gene": 0.4551,
        "Gene": "Apoe",
      },
      {
        "Average expression": 0.0449,
        "Cluster": "Cluster 5",
        "Fraction of cells expressing gene": 1.4031,
        "Gene": "Chil3",
      },
      {
        "Average expression": 0.0184,
        "Cluster": "Cluster 6",
        "Fraction of cells expressing gene": 0.5008,
        "Gene": "Chil3",
      },
      {
        "Average expression": 0.0206,
        "Cluster": "Cluster 7",
        "Fraction of cells expressing gene": 0.6211,
        "Gene": "Chil3",
      },
      {
        "Average expression": 0,
        "Cluster": "Cluster 8",
        "Fraction of cells expressing gene": 0,
        "Gene": "Chil3",
      },
      {
        "Average expression": 0.0073,
        "Cluster": "Cluster 9",
        "Fraction of cells expressing gene": 0.6173,
        "Gene": "Chil3",
      },
      {
        "Average expression": 0.0112,
        "Cluster": "Cluster 10",
        "Fraction of cells expressing gene": 0.7634,
        "Gene": "Chil3",
      },
      {
        "Average expression": 2.3613,
        "Cluster": "Cluster 11",
        "Fraction of cells expressing gene": 17.8571,
        "Gene": "Chil3",
      },
      {
        "Average expression": 10.7617,
        "Cluster": "Cluster 12",
        "Fraction of cells expressing gene": 37.5,
        "Gene": "Chil3",
      },
      {
        "Average expression": 0,
        "Cluster": "Cluster 13",
        "Fraction of cells expressing gene": 0,
        "Gene": "Chil3",
      },
      {
        "Average expression": 165.1493,
        "Cluster": "Cluster 14",
        "Fraction of cells expressing gene": 100,
        "Gene": "Chil3",
      },
      {
        "Average expression": 0.0227,
        "Cluster": "Cluster 0",
        "Fraction of cells expressing gene": 0.7174,
        "Gene": "Chil3",
      },
      {
        "Average expression": 0.0161,
        "Cluster": "Cluster 1",
        "Fraction of cells expressing gene": 0.7067,
        "Gene": "Chil3",
      },
      {
        "Average expression": 0.0297,
        "Cluster": "Cluster 2",
        "Fraction of cells expressing gene": 0.8447,
        "Gene": "Chil3",
      },
      {
        "Average expression": 0.0292,
        "Cluster": "Cluster 3",
        "Fraction of cells expressing gene": 0.7857,
        "Gene": "Chil3",
      },
      {
        "Average expression": 0.0407,
        "Cluster": "Cluster 4",
        "Fraction of cells expressing gene": 0.7964,
        "Gene": "Chil3",
      },
      {
        "Average expression": 0.3116,
        "Cluster": "Cluster 5",
        "Fraction of cells expressing gene": 5.102,
        "Gene": "Gzma",
      },
      {
        "Average expression": 20.4925,
        "Cluster": "Cluster 6",
        "Fraction of cells expressing gene": 23.5392,
        "Gene": "Gzma",
      },
      {
        "Average expression": 176.8351,
        "Cluster": "Cluster 7",
        "Fraction of cells expressing gene": 91.3043,
        "Gene": "Gzma",
      },
      {
        "Average expression": 115.7325,
        "Cluster": "Cluster 8",
        "Fraction of cells expressing gene": 66.9951,
        "Gene": "Gzma",
      },
      {
        "Average expression": 0.1699,
        "Cluster": "Cluster 9",
        "Fraction of cells expressing gene": 4.9383,
        "Gene": "Gzma",
      },
      {
        "Average expression": 0.4124,
        "Cluster": "Cluster 10",
        "Fraction of cells expressing gene": 6.8702,
        "Gene": "Gzma",
      },
      {
        "Average expression": 0.1222,
        "Cluster": "Cluster 11",
        "Fraction of cells expressing gene": 4.4643,
        "Gene": "Gzma",
      },
      {
        "Average expression": 0.1052,
        "Cluster": "Cluster 12",
        "Fraction of cells expressing gene": 2.2727,
        "Gene": "Gzma",
      },
      {
        "Average expression": 0.4625,
        "Cluster": "Cluster 13",
        "Fraction of cells expressing gene": 6.3492,
        "Gene": "Gzma",
      },
      {
        "Average expression": 0.6459,
        "Cluster": "Cluster 14",
        "Fraction of cells expressing gene": 11.7647,
        "Gene": "Gzma",
      },
      {
        "Average expression": 0.0839,
        "Cluster": "Cluster 0",
        "Fraction of cells expressing gene": 2.3434,
        "Gene": "Gzma",
      },
      {
        "Average expression": 0.2872,
        "Cluster": "Cluster 1",
        "Fraction of cells expressing gene": 3.7102,
        "Gene": "Gzma",
      },
      {
        "Average expression": 0.0892,
        "Cluster": "Cluster 2",
        "Fraction of cells expressing gene": 2.2742,
        "Gene": "Gzma",
      },
      {
        "Average expression": 0.3174,
        "Cluster": "Cluster 3",
        "Fraction of cells expressing gene": 5.7143,
        "Gene": "Gzma",
      },
      {
        "Average expression": 0.1183,
        "Cluster": "Cluster 4",
        "Fraction of cells expressing gene": 3.0717,
        "Gene": "Gzma",
      },
      {
        "Average expression": 0.0134,
        "Cluster": "Cluster 5",
        "Fraction of cells expressing gene": 0.3827,
        "Gene": "Ifitm3",
      },
      {
        "Average expression": 0.0311,
        "Cluster": "Cluster 6",
        "Fraction of cells expressing gene": 0.6678,
        "Gene": "Ifitm3",
      },
      {
        "Average expression": 0.0844,
        "Cluster": "Cluster 7",
        "Fraction of cells expressing gene": 0.8282,
        "Gene": "Ifitm3",
      },
      {
        "Average expression": 0.0659,
        "Cluster": "Cluster 8",
        "Fraction of cells expressing gene": 6.8966,
        "Gene": "Ifitm3",
      },
      {
        "Average expression": 0.2257,
        "Cluster": "Cluster 9",
        "Fraction of cells expressing gene": 3.0864,
        "Gene": "Ifitm3",
      },
      {
        "Average expression": 0.0667,
        "Cluster": "Cluster 10",
        "Fraction of cells expressing gene": 3.0534,
        "Gene": "Ifitm3",
      },
      {
        "Average expression": 43.5782,
        "Cluster": "Cluster 11",
        "Fraction of cells expressing gene": 99.1071,
        "Gene": "Ifitm3",
      },
      {
        "Average expression": 51.6382,
        "Cluster": "Cluster 12",
        "Fraction of cells expressing gene": 87.5,
        "Gene": "Ifitm3",
      },
      {
        "Average expression": 23.2764,
        "Cluster": "Cluster 13",
        "Fraction of cells expressing gene": 68.254,
        "Gene": "Ifitm3",
      },
      {
        "Average expression": 0.1597,
        "Cluster": "Cluster 14",
        "Fraction of cells expressing gene": 1.9608,
        "Gene": "Ifitm3",
      },
      {
        "Average expression": 0.0226,
        "Cluster": "Cluster 0",
        "Fraction of cells expressing gene": 0.7652,
        "Gene": "Ifitm3",
      },
      {
        "Average expression": 0.0255,
        "Cluster": "Cluster 1",
        "Fraction of cells expressing gene": 0.9423,
        "Gene": "Ifitm3",
      },
      {
        "Average expression": 0.0375,
        "Cluster": "Cluster 2",
        "Fraction of cells expressing gene": 1.0396,
        "Gene": "Ifitm3",
      },
      {
        "Average expression": 0.0238,
        "Cluster": "Cluster 3",
        "Fraction of cells expressing gene": 0.5,
        "Gene": "Ifitm3",
      },
      {
        "Average expression": 0.0275,
        "Cluster": "Cluster 4",
        "Fraction of cells expressing gene": 0.5688,
        "Gene": "Ifitm3",
      },
      {
        "Average expression": 0.0488,
        "Cluster": "Cluster 5",
        "Fraction of cells expressing gene": 1.5306,
        "Gene": "Lyz2",
      },
      {
        "Average expression": 0.0124,
        "Cluster": "Cluster 6",
        "Fraction of cells expressing gene": 0.5008,
        "Gene": "Lyz2",
      },
      {
        "Average expression": 0.0648,
        "Cluster": "Cluster 7",
        "Fraction of cells expressing gene": 1.8634,
        "Gene": "Lyz2",
      },
      {
        "Average expression": 0.034,
        "Cluster": "Cluster 8",
        "Fraction of cells expressing gene": 3.4483,
        "Gene": "Lyz2",
      },
      {
        "Average expression": 0.0565,
        "Cluster": "Cluster 9",
        "Fraction of cells expressing gene": 1.8519,
        "Gene": "Lyz2",
      },
      {
        "Average expression": 0.0293,
        "Cluster": "Cluster 10",
        "Fraction of cells expressing gene": 0.7634,
        "Gene": "Lyz2",
      },
      {
        "Average expression": 110.1555,
        "Cluster": "Cluster 11",
        "Fraction of cells expressing gene": 93.75,
        "Gene": "Lyz2",
      },
      {
        "Average expression": 238.1269,
        "Cluster": "Cluster 12",
        "Fraction of cells expressing gene": 86.3636,
        "Gene": "Lyz2",
      },
      {
        "Average expression": 2.088,
        "Cluster": "Cluster 13",
        "Fraction of cells expressing gene": 4.7619,
        "Gene": "Lyz2",
      },
      {
        "Average expression": 128.2302,
        "Cluster": "Cluster 14",
        "Fraction of cells expressing gene": 100,
        "Gene": "Lyz2",
      },
      {
        "Average expression": 0.0358,
        "Cluster": "Cluster 0",
        "Fraction of cells expressing gene": 1.1478,
        "Gene": "Lyz2",
      },
      {
        "Average expression": 0.0561,
        "Cluster": "Cluster 1",
        "Fraction of cells expressing gene": 2.1201,
        "Gene": "Lyz2",
      },
      {
        "Average expression": 0.0984,
        "Cluster": "Cluster 2",
        "Fraction of cells expressing gene": 2.3392,
        "Gene": "Lyz2",
      },
      {
        "Average expression": 0.054,
        "Cluster": "Cluster 3",
        "Fraction of cells expressing gene": 1.5,
        "Gene": "Lyz2",
      },
      {
        "Average expression": 0.1078,
        "Cluster": "Cluster 4",
        "Fraction of cells expressing gene": 1.3652,
        "Gene": "Lyz2",
      },
    ],
    "filename": "Test_Experiment-dot_plot-louvain-All",
  },
  {},
]
`;

exports[`Dot plot page Should show a no data error if user is using marker gene and selected filter sets are not represented in more than 1 group in the base cell set 1`] = `
[
  [
    "testae48e318dab9a1bd0bexperiment",
    {
      "limit": 100000,
      "name": "ListGenes",
      "offset": 0,
      "orderBy": "gene_names",
      "orderDirection": "ASC",
      "selectFields": [
        "gene_names",
        "dispersions",
      ],
    },
    [Function],
    [Function],
    {
      "timeout": 900,
    },
  ],
  [
    "testae48e318dab9a1bd0bexperiment",
    {
      "clusterNames": [
        "Cluster 0",
        "Cluster 1",
        "Cluster 2",
        "Cluster 3",
        "Cluster 4",
        "Copied WT2",
        "Copied KO",
        "WT2",
        "KO",
        "WT1",
      ],
      "customGenesList": [
        "Ccl5",
        "Chil3",
        "Lyz2",
      ],
      "filterBy": {
        "group": "All",
        "key": "All",
      },
      "groupBy": "louvain",
      "name": "DotPlot",
      "numberOfMarkers": 3,
      "useMarkerGenes": false,
    },
    [Function],
    [Function],
    {
      "timeout": 180,
    },
  ],
  [
    "testae48e318dab9a1bd0bexperiment",
    {
      "clusterNames": [
        "Cluster 0",
        "Cluster 1",
        "Cluster 2",
        "Cluster 3",
        "Cluster 4",
        "Copied WT2",
        "Copied KO",
        "WT2",
        "KO",
        "WT1",
      ],
      "customGenesList": [
        "Ccl5",
        "Chil3",
        "Lyz2",
      ],
      "filterBy": {
        "group": "All",
        "key": "All",
      },
      "groupBy": "louvain",
      "name": "DotPlot",
      "numberOfMarkers": 3,
      "useMarkerGenes": true,
    },
    [Function],
    [Function],
    {
      "timeout": 180,
    },
  ],
  [
    "testae48e318dab9a1bd0bexperiment",
    {
      "clusterNames": [
        "Cluster 0",
        "Cluster 1",
        "Cluster 2",
        "Cluster 3",
        "Cluster 4",
        "Copied WT2",
        "Copied KO",
        "WT2",
        "KO",
        "WT1",
      ],
      "customGenesList": [
        "Ccl5",
        "Chil3",
        "Lyz2",
      ],
      "filterBy": {
        "group": "All",
        "key": "All",
      },
      "groupBy": "sample",
      "name": "DotPlot",
      "numberOfMarkers": 3,
      "useMarkerGenes": true,
    },
    [Function],
    [Function],
    {
      "timeout": 180,
    },
  ],
]
`;
